support pure fp16 training in FSDP

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68417

1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

zhaojuanmao committed:
support pure fp16 training in FSDP

1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -391,11 +393,23 @@ def _init_param_attributes(self, p: Parameter) -> None:
                 )
             return
 
+        if hasattr(p, "_local_shard") and p.data != p._local_shard:  # type: ignore[attr-defined]
```
This seems like it should be an error? `p.data` is not the shard anymore?

zhaojuanmao replied:
it is possible that p.data is still pointing to sharded data, e..g, users convert the shard from fp16 to fp32 outside fsdp, but I agree that we should not support this use case, because users could do anything and we lose the check here

rohan-varma reviewed (commented):
> as parameters' data type may be changed after construction and before training loop

Is there a practical example of where this might happen in a real-world training loop?

zhaojuanmao commented:
> > as parameters' data type may be changed after construction and before training loop
> 
> Is there a practical example of where this might happen in a real-world training loop?

yes! That is how FairSeq stack is doing: wrapped a fp32 model, convert it to fp16 after wrapping and before training. Also feel it is good to get param's type during lazy_init where attributes start to initialize.

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_param_mutation.py:
```diff
@@ -0,0 +1,75 @@
+    def __init__(self):
+        super().__init__()
+        # keep everything deterministic for model initialization
+        torch.manual_seed(0)
```
nit: it doesn't need torch.cuda.manual_seed?

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -379,7 +377,15 @@ def _init_param_attributes(self, p: Parameter) -> None:
             p, "_orig_size"
         ), "Parameters should have been sharded during construction."
+        # If _local_shard has been set in the first lazy init and
+        # current parameter is pointed to _local_shard, no need to
```
👍

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -379,7 +377,15 @@ def _init_param_attributes(self, p: Parameter) -> None:
+        # set the _local_shard again.
         if hasattr(p, "_local_shard"):
+            assert p.data_ptr() == p._local_shard.data_ptr(), (  # type: ignore[attr-defined]
+                "Parameter storage is changed outside FSDP, this use case "
```
nit: I guess it can be changed outside of FSDP, but before first forward, i.e. we have the following in tests

model = FSDP(model)
model.half()
model(inputs)

but the issue is if its changed afeter the first iteration. Should we be more specific in this error message?

zhaojuanmao replied:
sounds good

rohan-varma reviewed (approved):
Amazing!

rohan-varma commented on test/distributed/fsdp/test_fsdp_pure_fp16.py:
```diff
@@ -0,0 +1,88 @@
+        # keep everything deterministic for input data
+        torch.manual_seed(0)
+
+        model = Model(wrap_fsdp).cuda()
```
should we test CPU offloading to ensure the data transfer between devices don't cause issues?

zhaojuanmao replied:
sounds good

rohan-varma commented on test/distributed/fsdp/test_fsdp_pure_fp16.py:
```diff
@@ -0,0 +1,88 @@
+
+        in_data = torch.rand(64, 4).cuda().half()
+        in_data.requires_grad = True
+        for _ in range(1):
```
run a couple iterations?

zhaojuanmao replied:
it will produce NaN value without grad scaler after 1st iteration for fp16 training..., with grad scaler, the DDP and FSDP results may not be exact the same then

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support pure fp16 training in FSDP"


1. since parameter attributes are lazily initialized at the beginning of forward, it makes more sense to init full_param_padded using parameters' data type during lazy_init, instead of using parameters' data type during construction, as parameters' data type may be changed after construction and before training loop
2.add checking whether parameter storage is changed outside FSDP and handle it properly

Differential Revision: [D32458643](https://our.internmc.facebook.com/intern/diff/D32458643/)

[ghstack-poisoned]
